src/VisualStudio/Core/Def/Implementation/Utilities/IVsShellExtensions.cs

```diff
@@ -3,12 +3,38 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.Utilities
 {
     internal static class IVsShellExtensions
     {
+        private static volatile int s_isInCommandLineMode;
+
+        public static bool IsInCommandLineMode
+        {
+            get
+            {
+                var result = s_isInCommandLineMode;
+                if (result == 0)
+                {
+                    s_isInCommandLineMode = result = ThreadHelper.JoinableTaskFactory.Run(async () =>
+                    {
+                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                        var shell = ServiceProvider.GlobalProvider.GetService<SVsShell, IVsShell>();
+                        return
+                            (shell != null) &&
+                            ErrorHandler.Succeeded(shell.GetProperty((int)__VSSPROPID.VSSPROPID_IsInCommandLineMode, out var result)) &&
+                            (bool)result ? 1 : -1;
```

Comment:
i'm worried about this throwing.  but i also don't know what the rigth thing is to do in that case.

Reply:
All call sites that use this API in devdiv just assume the value to be bool, so I think we are safe.